Document upgrades (#610)
weaveworks/eksctl: README.md

eksctl drain nodegroup --cluster=<clusterName> --name=<nodegroupName> --undo
```

### Cluster Upgrades

An _`eksctl`-managed_ cluster can be upgraded in 3 easy steps:

1. update control plane version with `eksctl update cluster`
2. replace each of the nodegroups by creating a new one and deleting the old one

> 3. update default add-ons:

Author: There a few objects there, we should just name the add-ons instead of referring to the workload objects...

  - `kube-system:ds/kube-proxy`
  - `kube-system:ds/aws-node`

Please make sure to read this section in full before you proceed.

> NOTE: Kubernetes supports version drift of up-to 2 minor versions during upgrade process.

### Updating control plane version

Control version updates must be done for one minor version at a time.

To update control plane to the next available version run:
```
eksctl update cluster --cluster=<clusterName>
```

This command will not apply any changes right away, you will need to re-run it with
`--dry-run=false` to apply the changes.

#### Updating nodegroups

You should update nodegroups only after you ran `eksctl update cluster`.

If you have a simple cluster with just an initial nodegroup (i.e. created with
`eksctl create cluster`), the process is very simple.

Get the name of old nodegroup:
```
eksctl get nodegroups --cluster=<clusterName>
```
> NOTE: you should see only one nodegroup here, if you see more - read the next section

Create new nodegroup:
```
eksctl create nodegroup --cluster=<clusterName>
```

Delete old nodegroup:
```
eksctl delete nodegroup --cluster=<clusterName> --name=<oldNodeGroupName>
```

> ##### Updating multiple nodegroups

Reviewer: do we want to split this section into `Updating multiple nodegroups with config file` and `Updating multiple nodegroups without config file`?

Author: Not sure at the moment, can always improve later on. We actually need a proper website for docs somewhat urgently.

If you have multiple nodegroups, it's your responsibility to track how each one was configured.
You can do this by using config files, but if you haven't used it already, you will need to inspect
your cluster to find out how each nodegroup was configured.

In general terms, you are looking to:
- review nodegroups you have and which ones can be delete or must be replaced with new version

> - note down configuration of each nodegroup, consider using config file to ease upgrades next time

Reviewer: is it obvious where to find the configuration for someone upgrading or do we need to give a hand here?

Author: Yes, we will need to, but cannot at the moment. We should move on to caching config inside the cluster, perhaps in a configmap to start with. This would need to be solved as we get closer to `eksctl apply` (also see #443).

To create a new nodegroup:
```
eksctl create nodegroup --cluster=<clusterName>
```

To delete old nodegroup:
```
eksctl create nodegroup --cluster=<clusterName>
```

If you are using config file, you will need to do the following.

Get the list of old nodegroups:
```

> old_nodegroups="$(eksctl get ng --cluster=<clusterName> --output=json | jq -r '.[].Name')"

Reviewer: why would i need to keep track of the old nodegroups when i use the config file to delete the old ones?

Author: I guess it all depends which way you edit the file and how you apply the edits, so you could delete old ones from the file to begin with avoid having to set `--only`. There is definitely much room for improvement of the UX overall, this one of the somewhat tricky things.

```

Edit config file to add new nodegroups.

> If you chose to deleted definition of old nodegoups from the config file, then you can run:

Reviewer: ```suggestion
If you choose to delete the definition of old nodegroups with the config file, then you can run:
```
(assuming here that nodegroups not found in the config file will be automatically deleted)

```

> eksctl create nodegroup --config-file=<path>

Reviewer: I'm not sure I understand this part, will this command delete nodegroups not in the config file?

Author: No, it will only create new ones, and unless you've deleted the definitions of the old ones before hand - it will complain about those. You'd want to set `--only=<listOfNewNGs>` otherwise.

Author: I think `--only` is documented quite clearly just below this.

```

Alternatively, you can pass `--only` with a list of the new nodegroup names that you want to

> add (you can use glob expressions, or list specific names - see above for details):

Reviewer: not sure where `see above` refers to

Author: Well spotted, will try to clarify.

```
eksctl create nodegroup --config-file=<path> --only=<list>
```

Once you have new nodegroups in place, you can delete old ones:
```
for ng in $old_nodegroups ; do eksctl delete nodegroup --cluster=<clusterName> --name=$ng ; done
```

#### Updating default add-ons

> There 3 default add-ons that get included in each EKS cluster, the process for updating each of them is different, hence

Reviewer: ```suggestion
There are 3 default add-ons that get included in each EKS cluster, the process for updating each of them is different, hence
```

there are 3 distinct commands that you will need to run.

To update `kube-system:ds/kube-proxy`, run:
```
eksctl utils update-kube-proxy
```

To update `kube-system:ds/aws-node`, run:
```
eksctl utils update-aws-node
```

> If you have upgraded from 1.10 to 1.11, you will need to replace `kube-system:deployment/kube-dns` with `kube-system:deployment/coredns`.

Reviewer: what if i'm going from 1.11 to 1.12, how do i upgrade `kube-dns`?

Author: That's unknown at present.

Create new nodegroup:
```
eksctl create nodegroup --cluster=<clusterName>
```

Delete old nodegroup:
```
eksctl delete nodegroup --cluster=<clusterName> --name=<oldNodeGroupName>
```

##### Updating multiple nodegroups

If you have multiple nodegroups, it's your responsibility to track how each one was configured.
You can do this by using config files, but if you haven't used it already, you will need to inspect
your cluster to find out how each nodegroup was configured.

In general terms, you are looking to:
- review nodegroups you have and which ones can be deleted or must be replaced for the new version
- note down configuration of each nodegroup, consider using config file to ease upgrades next time

To create a new nodegroup:
```
eksctl create nodegroup --cluster=<clusterName>
```

To delete old nodegroup:
```
eksctl create nodegroup --cluster=<clusterName>
```

##### Updating multiple nodegroups with config file

If you are using config file, you will need to do the following.

Get the list of old nodegroups:
```
old_nodegroups="$(eksctl get ng --cluster=<clusterName> --output=json | jq -r '.[].Name')"
```

Edit config file to add new nodegroups.

> If you choose to delete the definition of old nodegroups from the config file, then you can run:

Reviewer: i understood now, i think the `delete` made me think of `eksctl delete`, I would reword as such:
```suggestion
If you have removed the definition of the old nodegroups from the config file, then you can run:
```

```
eksctl create nodegroup --config-file=<path>
```

> Alternatively, you can keep the old definitions and pass `--only` with a list of the new nodegroup

Reviewer: ```suggestion
Alternatively, you can keep the old definitions in the config-file and pass `--only` with a list of the new nodegroup
```
